### DISCUSSION TITLE
/zones api with include_geometry=true doesn't return geometry

### DISCUSSION THREAD

**Discussion by dsadosky** [weather-gov/api#278] [Q&A]
The /zones api has a parameter to include_geometry which when set to true should return the zones geometry.  I am expecting it to return the geometry of the polygon(s) of the requested zone(s).  

However it appears that geometry is not returned in any case where include_geometry is set to true.

https://api.weather.gov/zones?id=MNC053&include_geometry=true

{
  "@context": {
    "@version": "1.1"
  },
  "type": "FeatureCollection",
  "features": [
    {
      "id": "https://api.weather.gov/zones/county/MNC053",
      "type": "Feature",
      **"geometry": null,**
      "properties": {
        "@id": "https://api.weather.gov/zones/county/MNC053",
        "@type": "wx:Zone",
        "id": "MNC053",
        "type": "county",
        "name": "Hennepin",
        "effectiveDate": "2019-05-02T18:00:00+00:00",
        "expirationDate": "2219-05-02T18:00:00+00:00",
        "state": "MN",
        "cwa": [
          "MPX"
        ],
        "forecastOffices": [
          "https://api.weather.gov/offices/MPX"
        ],
        "timeZone": [
          "America/Chicago"
        ],
        "observationStations": [],
        "radarStation": null
      }
    }
  ]
}

**Comment by JJKraw** [comment]
You probably know this, but I'll point it out just in case someone doesn't: If you use the endpoint of the form

  https://api.weather.gov/zones/county/MNC053

you will get the geometry as long as you request geo+json (which I believe is the default if you don't specify it in the request header).

**Comment by dsadosky** [comment]
@JJKraw Thanks.  Is the geometry returned at the forecast zone level or the entire county?  I would expect a similar response from the https://api.weather.gov/zones?id=MNC053&include_geometry=true but would expect the geometry to be at the forecast zone level.  Either way using include_geometry=true  with the zones api returns no geometry

**Reply by ephcoding** [reply]
@dsadosky if by chance you get/see this, you ever figure this out?  Seems like the <include_geometry> query param still doesn't work.  Making individual calls for each zone to get the geometry seems wicked expensive.

Thanks - E

**Reply by JJKraw** [reply]
Where the API offers an endpoint that returns exactly what I want, I avoid alternatives that use parameters. The example I gave above works.

**Reply by djdevdog** [reply]
I had a question on this, I'm doing something to receive alerts for the country as a whole and I noticed (at least just today) there were nearly 1000 unique zones listed (I'm sure it could be much higher on a given day). If I were to query this data from the API programmatically to get the geometry of each zone, how much time should I give between each request at minimum? I know the query/sec limit is not public knowledge for good reason and I saw it gives you 5 secs to query again if you do get booted.

**Reply by JJKraw** [reply]
First note: I use the API for a web site, so I have a Linux server in which I can store the geometry in cache files for the long term. So usually if I've retrieved a zone geometry once, I never have to fetch it again.

That said, I have a mechanism to refresh the files, and via trial and error I settled on 6 seconds by default, but for particularly large geometries, I use 30 seconds.  Those "long" codes are: "LAC075", "AKZ022", "AKZ023", "AKZ024", "AKZ025", "AKZ026", "AKZ027", "AKZ028", "AKZ029", "MEC009", "GMZ235", "GMZ552", "GMZ555", "GMZ850", "LAZ069", "NMZ401", "NMZ402", "PKZ031", "PKZ032", "PKZ034", "PKZ035",  "PKZ036", "PKZ041", "PKZ042", "PKZ043"

**Reply by dwhitemv25** [reply]
If it helps, you can download shapefiles for all NWS zones from:

https://www.weather.gov/gis/AWIPSShapefiles

The zones change very infrequently, and changes are announced via Service Change Notices (SCN), available here:

https://www.weather.gov/notification/

For instance, there are zone changes scheduled for March 5, 2024.

**Reply by sullynole** [reply]
There's also a more robust REST available for the zones maps here:  https://mapservices.weather.noaa.gov/static/rest/services/nws_reference_maps/nws_reference_map/MapServer

**Comment by JJKraw** [comment]
If you request a county (zones/county/xxCnnn), you get the county geometry. If you select a forecast zone (zones/forecast/xxZnnn), you get the zone geometry.

The amount of data returned can be huge, particularly if a coastline is involved.
